File: Packages/PassthroughCore/Sources/PassthroughCore/Stats.swift
import Foundation
import os

/// Lock-free-ish byte counters updated from network queues and read from the UI.
public final class ByteCounter: @unchecked Sendable {
    private let lock = OSAllocatedUnfairLock(initialState: (rx: Int64(0), tx: Int64(0), active: Int(0), total: Int(0)))

    public init() {}

    public func addRx(_ n: Int) { lock.withLock { $0.rx += Int64(n) } }
    public func addTx(_ n: Int) { lock.withLock { $0.tx += Int64(n) } }
    public func connectionOpened() { lock.withLock { $0.active += 1; $0.total += 1 } }
    public func connectionClosed() { lock.withLock { $0.active = max(0, $0.active - 1) } }

    public struct Snapshot: Sendable, Equatable {
        public var rx: Int64
        public var tx: Int64
        public var active: Int
        public var totalConnections: Int
        public init(rx: Int64, tx: Int64, active: Int, totalConnections: Int) {
            self.rx = rx; self.tx = tx; self.active = active; self.totalConnections = totalConnections
        }
        public static let zero = Snapshot(rx: 0, tx: 0, active: 0, totalConnections: 0)
    }

    public func snapshot() -> Snapshot {
        lock.withLock { Snapshot(rx: $0.rx, tx: $0.tx, active: $0.active, totalConnections: $0.total) }
    }

    public func reset() { lock.withLock { $0 = (0, 0, $0.active, 0) } }
}

/// One second of throughput, used for sparkline charts.
public struct ThroughputSample: Identifiable, Sendable, Equatable {
    public let id: Int
    public let date: Date
    /// Bytes per second.
    public let down: Double
    public let up: Double
    public init(id: Int, date: Date, down: Double, up: Double) {
        self.id = id; self.date = date; self.down = down; self.up = up
    }
}

/// Turns raw byte counters into rates and a short history window.
public struct TrafficMeter: Sendable, Equatable {
    public private(set) var history: [ThroughputSample] = []
    public private(set) var downRate: Double = 0
    public private(set) var upRate: Double = 0
    public private(set) var last: ByteCounter.Snapshot = .zero
    public private(set) var lastDate: Date?
    public let windowSize: Int
    private var counter = 0

    public init(windowSize: Int = 60) {
        self.windowSize = windowSize
        let now = Date()
        history = (0..<windowSize).map { i in
            ThroughputSample(id: i - windowSize, date: now.addingTimeInterval(Double(i - windowSize)), down: 0, up: 0)
        }
        counter = 0
    }

    public mutating func record(_ snapshot: ByteCounter.Snapshot, at date: Date = Date()) {
        defer { last = snapshot; lastDate = date }
        guard let lastDate else { return }
        let dt = max(0.05, date.timeIntervalSince(lastDate))
        downRate = max(0, Double(snapshot.rx - last.rx) / dt)
        upRate = max(0, Double(snapshot.tx - last.tx) / dt)
        history.append(ThroughputSample(id: counter, date: date, down: downRate, up: upRate))
        counter += 1
        if history.count > windowSize { history.removeFirst(history.count - windowSize) }
    }

    public mutating func reset() {
        self = TrafficMeter(windowSize: windowSize)
    }

    public var peakRate: Double {
        history.reduce(0) { max($0, max($1.down, $1.up)) }
    }
}

public enum ByteFormat {
    public static func bytes(_ value: Int64) -> String {
        let f = ByteCountFormatter()
        f.countStyle = .binary
        f.allowsNonnumericFormatting = false
        return f.string(fromByteCount: value)
    }

    /// e.g. "12.4 MB/s"; always two significant components for stable layout.
    public static func rate(_ bytesPerSecond: Double) -> (value: String, unit: String) {
        let units = ["B/s", "KB/s", "MB/s", "GB/s"]
        var v = bytesPerSecond
        var i = 0
        while v >= 1000 && i < units.count - 1 { v /= 1000; i += 1 }
        let text = i == 0 ? String(format: "%.0f", v) : (v < 10 ? String(format: "%.2f", v) : (v < 100 ? String(format: "%.1f", v) : String(format: "%.0f", v)))
        return (text, units[i])
    }

    public static func duration(_ seconds: TimeInterval) -> String {
        let s = Int(seconds)
        if s < 3600 { return String(format: "%02d:%02d", s / 60, s % 60) }
        return String(format: "%d:%02d:%02d", s / 3600, (s % 3600) / 60, s % 60)
    }
}

File: android/app/src/main/java/dev/dpatel/passthrough/core/Stats.kt
package dev.dpatel.passthrough.core

import java.util.Locale
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/** Byte counters updated from network threads and read from the UI. */
class ByteCounter {
    private val rx = AtomicLong()
    private val tx = AtomicLong()
    private val active = AtomicInteger()
    private val total = AtomicInteger()

    fun addRx(n: Int) { rx.addAndGet(n.toLong()) }
    fun addTx(n: Int) { tx.addAndGet(n.toLong()) }
    fun connectionOpened() { active.incrementAndGet(); total.incrementAndGet() }
    fun connectionClosed() { active.updateAndGet { maxOf(0, it - 1) } }

    data class Snapshot(val rx: Long, val tx: Long, val active: Int, val totalConnections: Int) {
        companion object { val ZERO = Snapshot(0, 0, 0, 0) }
    }

    fun snapshot() = Snapshot(rx.get(), tx.get(), active.get(), total.get())
}

/** One second of throughput, used for the chart. */
data class ThroughputSample(val id: Int, val date: Long, val down: Double, val up: Double)

/** Turns raw byte counters into rates and a short history window. */
data class TrafficMeter(
    val history: List<ThroughputSample> = emptyHistory(60),
    val downRate: Double = 0.0,
    val upRate: Double = 0.0,
    val last: ByteCounter.Snapshot = ByteCounter.Snapshot.ZERO,
    val lastDate: Long? = null,
    val windowSize: Int = 60,
    private val counter: Int = 0,
) {
    fun record(snapshot: ByteCounter.Snapshot, date: Long = System.currentTimeMillis()): TrafficMeter {
        val previous = lastDate ?: return copy(last = snapshot, lastDate = date)
        val dt = maxOf(0.05, (date - previous) / 1000.0)
        val down = maxOf(0.0, (snapshot.rx - last.rx) / dt)
        val up = maxOf(0.0, (snapshot.tx - last.tx) / dt)
        val h = (history + ThroughputSample(counter, date, down, up)).takeLast(windowSize)
        return copy(history = h, downRate = down, upRate = up, last = snapshot, lastDate = date, counter = counter + 1)
    }

    val peakRate: Double get() = history.maxOfOrNull { maxOf(it.down, it.up) } ?: 0.0

    companion object {
        fun emptyHistory(n: Int): List<ThroughputSample> {
            val now = System.currentTimeMillis()
            return (0 until n).map { i -> ThroughputSample(i - n, now + (i - n) * 1000L, 0.0, 0.0) }
        }
    }
}

object ByteFormat {
    /** Binary units with the same wording as Apple's ByteCountFormatter (.binary). */
    fun bytes(value: Long): String {
        if (value < 1024) return "$value bytes"
        val units = listOf("KB", "MB", "GB", "TB")
        var v = value / 1024.0
        var i = 0
        while (v >= 1024 && i < units.size - 1) { v /= 1024; i++ }
        val text = if (v < 10 && i > 0) String.format(Locale.US, "%.2f", v) else if (v < 100) String.format(Locale.US, "%.1f", v) else String.format(Locale.US, "%.0f", v)
        return "$text ${units[i]}"
    }

    /** e.g. "12.4" + "MB/s"; always two significant components for stable layout. */
    fun rate(bytesPerSecond: Double): Pair<String, String> {
        val units = listOf("B/s", "KB/s", "MB/s", "GB/s")
        var v = bytesPerSecond
        var i = 0
        while (v >= 1000 && i < units.size - 1) { v /= 1000; i++ }
        val text = when {
            i == 0 -> String.format(Locale.US, "%.0f", v)
            v < 10 -> String.format(Locale.US, "%.2f", v)
            v < 100 -> String.format(Locale.US, "%.1f", v)
            else -> String.format(Locale.US, "%.0f", v)
        }
        return text to units[i]
    }

    fun duration(seconds: Long): String {
        val s = maxOf(0, seconds)
        return if (s < 3600) String.format(Locale.US, "%02d:%02d", s / 60, s % 60)
        else String.format(Locale.US, "%d:%02d:%02d", s / 3600, (s % 3600) / 60, s % 60)
    }
}
